contracts/OwnerData.sol

```diff
@@ -0,0 +1,241 @@
+            result[i] = temp[i];
+        }
+        return result;
+    }
+
+    /// @notice remove data by indexes
+    /// @param contractAddress_ - the address of the contract
+    /// @param tokenID_ - the token ID
+    /// @param indexes_ - the indexes of the data to remove
+    function remove(address contractAddress_, uint256 tokenID_, uint256[] calldata indexes_) external onlyOwner {
+        Data[] storage data = _tokenData[contractAddress_][tokenID_];
+        for (uint256 i = 0; i < indexes_.length; i++) {
+            data[indexes_[i]].dataHash = new bytes(0);
+        }
+        emit DataRemoved(contractAddress_, tokenID_, indexes_);
+    }
+
+    /// @notice the service fee of adding data
+    /// @param serviceFee_ - the service fee of adding data
+    function setServiceFee(uint256 serviceFee_) external onlyOwner {
+        serviceFee = serviceFee_;
+    }
+
+    /// @notice the service fee receiver address
+    /// @param serviceFeeReceiver_ - the address of service fee receiver
+    function setServiceFeeReceiver(address serviceFeeReceiver_) external onlyOwner {
+        if (serviceFeeReceiver_ == address(0)) {
+            revert EmptyServiceFeeReceiver();
+        }
+        serviceFeeReceiver = serviceFeeReceiver_;
+    }
+
+    /// @notice the address of the signer
+    /// @param signer_ - the address of the signer
+    function setSigner(address signer_) external onlyOwner {
+        if (signer_ == address(0)) {
+            revert TrusteeIsZeroAddress();
+        }
+        signer = signer_;
+    }
+
+    /// @notice set public tokens
+    /// @param contractAddresses_ - the addresses of the contracts
+    /// @param tokenIDs_ - the token IDs
+    /// @param isPublic_ - the flag of public token
+    function setPublicTokens(address[] memory contractAddresses_, uint256[] memory tokenIDs_, bool isPublic_) external onlyOwner {
+        if (contractAddresses_.length == 0 || contractAddresses_.length != tokenIDs_.length) {
+            revert InvalidParameters();
+        }
+        for (uint256 i = 0; i < contractAddresses_.length; i++) {
+            _publicTokens[contractAddresses_[i]][tokenIDs_[i]] = isPublic_;
+        }
+    }
+
+    /// @notice add data with signature
+    /// @param contractAddress_ - the address of the contract
+    /// @param tokenID_ - the token ID
+    /// @param data_ - the data to add
+    /// @param signature_ - the signature
+    function signedAdd(address contractAddress_, uint256 tokenID_, Data calldata data_, Signature calldata signature_) external {
+        _validateSignature(signature_);
+        address account = data_.owner;
+        if (!_publicTokens[contractAddress_][tokenID_]) {
+            account = _recoverOwnerSignature(signature_.ownerSign);
+        }
+        _addData(account, contractAddress_, tokenID_, data_);
+    }
+
+    function _addData(address sender_, address contractAddress_, uint256 tokenID_, Data memory data_) private {
```

Comment:
Add `bool isPublicToken_` so we don't have to read from storage again.

Reply:
But we still need to read and verify, no?
Or we can set it immutable.